### ISSUE TITLE
Enable HEVC encoding

### ISSUE THREAD

**Issue by startergo** [acidanthera/bugtracker#668]
On a cMP5,1 with Westmere processors and RX-580 video cards HEVC encoding is achieved
by only adding:
` <key>PlatformInfo</key>
    <dict>
        <key>Automatic</key>
        <false/>
        <key>SMBIOS</key>
        <dict>
            <key>BoardProduct</key>
            <string>Mac-7BA5B2D9E42DDD94</string>
        </dict>
        <key>UpdateDataHub</key>
        <false/>
        <key>UpdateNVRAM</key>
        <false/>
        <key>UpdateSMBIOS</key>
        <true/>
        <key>UpdateSMBIOSMode</key>
        <string>Create</string> `
Adding these boot arguments complements the above with full DRM playback support:
`-wegtree shikigva=80`
Apparently only 
`<key>SMBIOS</key>
        <dict>
            <key>BoardProduct</key>
            <string>Mac-7BA5B2D9E42DDD94</string>`
Triggers HEVC encoding. Is there a possibility to integrate this in the WEG rather than changing the SMBIOS as there are some undesired changes to the CPU power management if SMBIOS is modified on that platform?

**Comment by cattyhouse** [comment]
does `shikigva=32 shiki-id=Mac-7BA5B2D9E42DDD94` work for you?

**Comment by startergo** [comment]
It will not acheive HEVC encoding nor DRM playback.

**Comment by Andrey1970AppleLife** [comment]
https://github.com/acidanthera/bugtracker/issues/553
-wegtree shikigva=32 shiki-id=Mac-7BA5B2D9E42DDD94

**Comment by startergo** [comment]
Andrey have you tested this on the same platform with the same settings?
With your settings HEVC uses the CPU instead of the GPU and the DRM playback does not work, so yes it is a valid issue. With the settings described in the beginning the HEVC encoding is performed by the GPU not the CPU and one can play all DRM contents on Safari in Netflix, Prime video, Apple TV, etc.

**Comment by Andrey1970AppleLife** [comment]
Enable bit32 + bit64, shikigva=96

**Comment by startergo** [comment]
Andrey first of all thank you for the support. I have been testing this for several months and with every WEG release. For instance in the previous release 144 + imacpro board_id in SMBIOS achieved this.
In the last release 80+imacpro board-id in the SMBIOS works. I see you repurposed 80 and enabled higher DRM support. I have tested all possible combinations and only those 2 acheived everything. I just recently tested 16,32,60, 96, 104,144, 160, 192 etc... 
So as you can see I am really puzzled to why only imacpro board_id in SMBIOS enables HEVC hardware encoding (GPU)? There is something else which is missing in the  WEG patches if this is the case.

**Comment by vit9696** [comment]
@startergo HEVC **encoder** enabling has never been done in WEG. The patch responsible for board-id swap is [here](https://github.com/acidanthera/WhateverGreen/blob/1ac7d71/Resources/Patches.plist#L224-L241). It is almost certainly the same for HEVC, but one needs to find a framework to add it to.

You can use [`ack`](https://github.com/beyondgrep/ack3) to find `board-id` in binary files in `/System`. Like this:

```
ack -l --type-set='all:match:.*' -k board-id /System
```

Then it is trial and error to find the right framework. Once you find it, let us know and we will reopen this bugreport and/or add the patches to WEG. Please note, that HEVC [decoder](https://github.com/acidanthera/WhateverGreen/blob/1ac7d71/Resources/Patches.plist#L338-L343)/encoder (not present in WEG, but should have similar paths) are XPCs. Depending on the framework/library one may need to explicitly specify the file in Patches.plist. If you are unsure, you could always add encoder XPC the same way decoder is added, and enable whitelist bit.

**Comment by startergo** [comment]
> If you are unsure, you could always add encoder XPC the same way decoder is added, and enable whitelist bit.

Forgive my ignorance, but isn't WHITELIST enabled with shikigva=8? In this case I have 64+16 (HWDRMID) or 128+16 enabling DRM playback on the top of HEVC encode/decode acheived solely by SMBIOS iMacPro1,1 board-id spoofing in the SMBIOS.
 So should I enable bit WHITELIST or HWDRMID (or whichever gives me 128+16)?
What is NSTREAM? is it UseHwDrmStreaming?

**Comment by startergo** [comment]
So these are the processes which are used in HEVC decoding through the AMD:
Code:
```
/System/Library/Frameworks/VideoToolbox.framework/Versions/A/XPCServices/VTDecoderXPCService.xpc/Contents/MacOS/VTDecoderXPCService
/System/Library/Video/Plug-Ins/VCPHEVC.bundle/Contents/MacOS/VCPHEVC
/System/Library/Video/Plug-Ins/AppleGVAHEVCDecoder.bundle/Contents/MacOS/AppleGVAHEVCDecoder
/System/Library/Extensions/AMDRadeonVADriver2.bundle/Contents/MacOS/AMDRadeonVADriver2
/System/Library/Frameworks/VideoToolbox.framework/Versions/A/Resources/default.metallib
```
And these are the processes used in HEVC encoding WITHOUT the GPU:
Code:

```
/System/Library/Frameworks/VideoToolbox.framework/Versions/A/XPCServices/VTEncoderXPCService.xpc/Contents/MacOS/VTEncoderXPCService
/System/Library/Video/Plug-Ins/AppleGVAHEVCEncoder.bundle/Contents/MacOS/AppleGVAHEVCEncoder
/System/Library/Video/Plug-Ins/VCPHEVC.bundle/Contents/MacOS/VCPHEVC
```

When iMacPro1,1 in the SMBIOS is spoofed we get HEVC encoding through the GPU
Code
```
/System/Library/Frameworks/VideoToolbox.framework/Versions/A/XPCServices/VTEncoderXPCService.xpc/Contents/MacOS/VTEncoderXPCService
/System/Library/Video/Plug-Ins/AppleGVAHEVCEncoder.bundle/Contents/MacOS/AppleGVAHEVCEncoder
/System/Library/Extensions/AMDRadeonVADriver2.bundle/Contents/MacOS/AMDRadeonVADriver2
/System/Library/Video/Plug-Ins/AppleGVAHEVCEncoder.bundle/Contents/Resources/AppleGVAHEVCFrameStatistics.metallib
```
Netflix Hardware DRM playback in Safari:
Code
```
/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc
/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/SystemAppearance.car
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/VibrantLightAppearance.car
/System/Library/PrivateFrameworks/DataDetectorsCore.framework/Versions/A/PlugIns/PhoneNumbers.plugin/Versions/A/Resources/main.cache
/System/Library/Extensions/AppleHDA.kext/Contents/PlugIns/AppleHDAHALPlugIn.bundle/Contents/MacOS/AppleHDAHALPlugIn
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/DarkAquaAppearance.car
/System/Library/Extensions/AMDRadeonVADriver2.bundle/Contents/MacOS/AMDRadeonVADriver2
/System/Library/PrivateFrameworks/CoreLSKDMSE.framework/Versions/A/CoreLSKDMSE
/System/Library/Extensions/AMDRadeonX4000GLDriver.bundle/Contents/MacOS/AMDRadeonX4000GLDriver
/System/Library/Components/AudioCodecs.component/Contents/MacOS/AudioCodecs
/System/Library/Components/CoreAudio.component/Contents/MacOS/CoreAudio
/System/Library/PrivateFrameworks/CoreKE.framework/Versions/A/CoreKE
/System/Library/Frameworks/QuartzCore.framework/Versions/A/Resources/default.metallib
```
Prime Video Hardware DRM Playback:
```
/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc
/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/SystemAppearance.car
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/VibrantLightAppearance.car
/System/Library/Extensions/AppleHDA.kext/Contents/PlugIns/AppleHDAHALPlugIn.bundle/Contents/MacOS/AppleHDAHALPlugIn
/System/Library/PrivateFrameworks/DataDetectorsCore.framework/Versions/A/PlugIns/PhoneNumbers.plugin/Versions/A/Resources/main.cache
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/DarkAquaAppearance.car
/System/Library/Components/AudioCodecs.component/Contents/MacOS/AudioCodecs
/System/Library/Components/CoreAudio.component/Contents/MacOS/CoreAudio
/System/Library/Extensions/AMDRadeonVADriver2.bundle/Contents/MacOS/AMDRadeonVADriver2
/System/Library/Extensions/AMDRadeonX4000GLDriver.bundle/Contents/MacOS/AMDRadeonX4000GLDriver
/System/Library/PrivateFrameworks/CoreKE.framework/Versions/A/CoreKE
/System/Library/PrivateFrameworks/CoreLSKD.framework/Versions/A/CoreLSKD
/System/Library/Frameworks/QuartzCore.framework/Versions/A/Resources/default.metallib
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_stdlib.metallib
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_filters.metallib
```
Apple TV Hardware DRM playback:
Code
```
/System/Applications/TV.app/Contents/MacOS/TV
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/SystemAppearance.car
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/DarkAquaAppearance.car
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/DarkAppearance.car
/System/Library/PrivateFrameworks/CoreFP.framework/Versions/A/CoreFP
/System/Library/Extensions/AppleHDA.kext/Contents/PlugIns/AppleHDAHALPlugIn.bundle/Contents/MacOS/AppleHDAHALPlugIn
/System/Applications/TV.app/Contents/Resources/Assets.car
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/Assets.car
/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app/Contents/PlugIns/SimulatorShareExtension.appex/Contents/Resources/Assets.car
/System/Applications/Reminders.app/Contents/PlugIns/RemindersSharingExtension.appex/Contents/Resources/Assets.car
/System/Library/PrivateFrameworks/FinderKit.framework/Versions/A/PlugIns/AirDrop.appex/Contents/Resources/Assets.car
/System/Library/PrivateFrameworks/ShareKit.framework/Versions/A/Resources/Assets.car
/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/Resources/Extras2.rsrc
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/VibrantLightAppearance.car
/System/Library/Keyboard Layouts/AppleKeyboardLayouts.bundle/Contents/Resources/AppleKeyboardLayouts-L.dat
/System/Library/Caches/com.apple.IntlDataCache.le.kbdx
/System/Library/Extensions/AMDRadeonX4000GLDriver.bundle/Contents/MacOS/AMDRadeonX4000GLDriver
/System/Library/PrivateFrameworks/MarkupUI.framework/PlugIns/Markup.appex/Contents/Resources/Assets.car
/System/Library/Frameworks/CoreMediaIO.framework/Versions/A/Resources/VDC.plugin/Contents/MacOS/VDC
/System/Library/PrivateFrameworks/CoreLSKD.framework/Versions/A/CoreLSKD
/System/Library/Extensions/AMDRadeonVADriver2.bundle/Contents/MacOS/AMDRadeonVADriver2
/System/Library/PrivateFrameworks/CoreKE.framework/Versions/A/CoreKE
/System/Library/Components/AudioCodecs.component/Contents/MacOS/AudioCodecs
/System/Library/Components/CoreAudio.component/Contents/MacOS/CoreAudio
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_stdlib.metallib
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_filters.metallib
```

**Comment by startergo** [comment]
Adding this Patch enables AppleTV+ DRM playback with "shikigva=80 -wegtree". No SMBIOS spoofing needed:
```
<dict>
                    <key>CPU</key>
                    <string>CPU_TYPE_X86_64</string>
                    <key>Comment</key>
                    <string>mov rax,1 : cpuid -&gt;  mov rax,306C3h : nop : nop (i7 4770k), allow hardware decoding for Netflix on models older than Haswell</string>
                    <key>Count</key>
                    <integer>2</integer>
                    <key>Find</key>
                    <data>
                    x8ABAAAAD6I=
                    </data>
                    <key>Replace</key>
                    <data>
                    x8DDBgMAkJA=
                    </data>
                    <key>Section</key>
                    <string>HWDRMID</string>
                    <key>Skip</key>
                    <integer>0</integer>
                </dict>
```

**Comment by vit9696** [comment]
Ugh, which section did you add it to? This patch is already applied by default with shikigva=16 or shikigva=80 (but shikigva=16 is enough for AppleTV).

**Comment by startergo** [comment]
> Ugh, which section did you add it to? This patch is already applied by default with shikigva=16 or shikigva=80 (but shikigva=16 is enough for AppleTV).

```
/System/Library/PrivateFrameworks/CoreLSKDMSE.framework/Versions/A/CoreLSKDMSE
```
Without SMBIOS board-id spoofed to iMac Pro only HEVC decoding worked (partially I did some other patches which enhaced HEVC decode in other apps)
So I tried different patches for HWDRMID (16+64). This particular one enabled Apple TV only for some reason, but I can't get HEVC hardware nor Netflix, Primevideo DRM without smbios spoofing. My goal is to keep the smbios untouched (original SMBIOS) and everything else working just like it does with HWDRMID and smbios spoofing. Maybe I need to patch binary 
```
/System/Library/PrivateFrameworks/AppleGVA.framework/Versions/A/AppleGVA
```
from
7BA5B2D9E42DDD94 to F221BEC8? And all other relevant frameworks and files?
What would be a binary patch template? Can it be done in the Kernel-->Patch section of OC?
